pkg/gatewayserver/gatewayserver.go

```diff
@@ -664,12 +664,13 @@ func (gs *GatewayServer) handleUpstream(conn connectionEntry) {
 						pass = true
 					}
 					if !pass {
-						drop(ids, errNoRoute.WithAttributes("host", host.name))
 						break
 					}
-					if err := host.handler.HandleUplink(ctx, gtw.GatewayIdentifiers, ids, msg); err != nil {
+					switch err := host.handler.HandleUplink(ctx, gtw.GatewayIdentifiers, ids, msg); codes.Code(errors.Code(err)) {
+					case codes.Canceled, codes.DeadlineExceeded,
+						codes.Unknown, codes.Internal,
+						codes.Unimplemented, codes.Unavailable:
 						drop(ids, errHostHandle.WithCause(err).WithAttributes("host", host.name))
-						break
 					}
 					registerForwardUplink(ctx, gtw, msg.UplinkMessage, host.name)
```

Comment:
So with for instance `DeadlineExceeded`, there's a drop and a forward registration. That doesn't seem right. Shouldn't `registerForwardUplink()` be called if the error code is `OK`?

Reply:
This should have been in the `default` case; fixed that.